book-3-web-api/chapters/FIREBASE_AUTH.md

```diff
@@ -0,0 +1,197 @@
+
+https://github.com/nashville-software-school/WisdomAndGrace
+
+### Look around the repo
+
+In the repo you will find...
+
+* A Visual Studio solution and project both called `WisdomAndGrace`.
+* A `client` folder inside the project folder than contains a react application. This application's `package.json` file has already been configured to use the react proxy server.
+* A sql script for building the application database.
+* An `appsettings.json` file containing configuration information for the Web API.
+* A `.env` file containing configuration settings for the React application.
+
+### Update the Repo
+
+Before we can run the app we need to make a few changes.
+
+1. Open the SQL script in an editor and find the `INSERT` statements that insert records into the `UserProfile` table. Modify these statements to add the users you added to Firebase earlier in this chapter.
+1. Update the `appsettings.json` file. Change the value of the `FirebaseProjectId` key to me the Project Id of your Firebase project.
+1. Update the `.env` file. Change the value of the `REACT_APP_API_KEY` key to the API Key from your Firebase project.
+
+Now we're finally ready to run this thing. Build your database, then run both server and client apps.
+
+## Examining the Demo
+
+Your instructor will take you through the parts of the demo application and describe how it is setup to use Firebase authentication.
+
+### Some Key Areas to Explore
+
+* JWT Authentication in `Startup.cs`.
+* `app.UseAuthentication()` in `Startup.cs`.
+* `FirebaseUserId` property in `UserPofile.cs`.
+* `GetCurrentUser()` method in `QuoteController`.
+* Use of the `[Authorize]` attribute to secure the `UserProfileController` and `QuoteController` controllers.
+* Firebase JavaScript Library installed from npm.
+* `Login`, `Register` and `Logout` functions in `UserProfileProvider.js`.
+* Using the token (JWT) in the `fetch()` calls in both `UserProfileProvider.js` and `QuoteProvider.js`.
+* Simple `UserType` verification in `QuoteController.cs` and in `QuoteProvider.js`.
+
+## Exercise
+
+1. Go through the chapter and follow the steps to make the `WisdomAndGrace` app work on your computer.
+1. Add Firebase authentication to the Gifter application.
+
+### Firebase Authentication Checklist
+
+#### In Firebase
+
+1. Create a new Firebase project.
+1. Enable the `Email/Password` `Sign-in method`.
+1. Add at least two new users to Firebase. Each user **must** have the same email address as one of the users in your Gifter application.
+
+#### In the Gifter Database
+
+> **NOTE:** The following will delete and recreate the database.
+
+1. Update the Gifter SQL script to add a `FirebaseUserId` column to the `UserProfile` table.
+
+    ```sql
+    CREATE TABLE [UserProfile] (
+    [Id] INTEGER PRIMARY KEY IDENTITY NOT NULL,
+    [FirebaseUserId] NVARCHAR(28) NOT NULL,
```

Comment:
I don't think it matters much, but can we add a `UNIQUE` here so we get sqlserver to add an index?